### PR TITLE
Accept AtomGroup alongwith Universe in the DistanceMatrix

### LINKED ISSUE #3486: DistanceMatrix accept AtomGroup
## Is your feature request related to a problem? ##

Is there any reason DistanceMatrix cannot accept an AtomGroup inplace of a Universe? This would help with atomgroups created through non-selection-string ways, which you then need to re-create with atom indices. e.g.
https://groups.google.com/g/mdnalysis-discussion/c/ad4HW_6w2MM/m/-h3bRDv0DgAJ?utm_medium=email&utm_source=footer 

## Describe the solution you'd like ##

Here: https://github.com/MDAnalysis/mdanalysis/blob/5bf6359a821c085a1039ca414ce51ca2820722fb/package/MDAnalysis/analysis/diffusionmap.py#L251-L253

```diff
- universe.trajectory
+ universe.universe.trajectory
```

### PR DESCRIPTION
**Pull request by AnirG**
Fixes #3486

Changes made in this Pull Request:
 - Accepts `AtomGroup` instead of `Universe` in the DistanceMatrix.
 - Modified tests to account for the modified input parameter.  

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (richardjgowers) [inline package/MDAnalysis/analysis/diffusionmap.py]: This is missing the point of the change. Currently a Universe and selection string are both passed to construct an AtomGroup. The desired change is for this selection to happen outside of the function, so these two arguments are replaced by just an AtomGroup. To not break things abruptly, this will therefore require for a while accepting either U+selection (and warning about the change) OR AG.

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py]: I see. I thought we'd have AG along with `selection` anyways should the need arise. I'll make the necessary changes to accept AG along with `Universe` and `selection`.

**Comment by AnirG** [comment]
@richardjgowers does it seem fine? please let me know if I should add any more tests or any modification is required.

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: For *new* tests we may as well use `assert_allclose` if the matrices contain floats per https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_almost_equal.html

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: can you briefly explain what this change is for? The `u` fixture returns a universe and now we need another `.universe` call?

Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: yeah, since `u` can also be an `AtomGroup`.  I should've also updated `match` as `match='U is not a Universe or AtomGroup'`. I will do that.

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: sure, will update it.

Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: Maybe we should consider replacing `assert_array_almost_equal` in other places as well as per the above link. I've updated it for only this test function.

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: At this moment in time our approach is to only change tests we touch and make sure new tests use assert_allclose. Making large blanket changes makes it very hard to review PRs and also understand why changes were made based on history. If we need to do a blanket change we will do this in a specific PR in the future.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (richardjgowers) [inline package/MDAnalysis/analysis/diffusionmap.py:248]: This change still needs documenting in the docstring for the class.

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: @tylerjereddy is correct here, u.trajectory == u.universe.trajectory, there's no need to make this change, either way you are testing what happens if you pass a trajectory through (neither a universe or an atomgroup)

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py:75]: Honestly as much as I like fixtures I'm not sure it's worth building a module-level fixture just for the one test, particularly when this can just be `u.select_atoms('backbone')`

**Review by IAlibay — CHANGES_REQUESTED** [review]
thanks @AnirG just a couple of small additional comments (on top of the doc changes Richard asked for).

**Comment by AnirG** [comment]
Hi, I'm not sure as to why the one codecov test failed.

**Comment by AnirG** [comment]
@tylerjereddy Hi looks like even #3527 failed similar tests on macOS like I'm facing. How should that be taken care of?

**Comment by IAlibay** [comment]
This is related to https://github.com/MDAnalysis/mdanalysis/issues/3550 and some further issues upstream - we are currently dealing with this, there is nothing to do about this in this specific PR.

**Comment by AnirG** [comment]
Hi @IAlibay! Can you please suggest any more changes? I'm looking forward to taking up more issues once I get done with this one, and possibly would love to hear some general practices I might've missed here.

**Comment by AnirG** [comment]
@IAlibay Hi! Do the final changes seem fine? Would make the necessary changes otherwise.

**Review by jbarnoud — CHANGES_REQUESTED** [review]
I restarted the tests as they seem to have failed on an unrelated issue. Assuming the tests do pass now, I think all the comments have been addressed appropriately. Thank you.

A couple of things are missing:
* [ ] you need to update the CHANGELOG
* [ ] you need to add yourself to the AUTHORS file

We hadn't release the 2.1.0 version when you last modified your pull request, so you will need to merge the develop branch or rebase against it to edit the most up to date version of these files.

Also, could you introduce yourself on the mailing list? We are discussing changes that will require us to contact all the contributors so we try to make sure we know how to reach new contributors. Thank you. (If you are interested in the [Google Summer of Code](https://www.mdanalysis.org/2022/03/07/gsoc2022/), please say so explicitly; many people are introducing themselves at the moment in that context and we need to know who's a candidate and who's not :smiley:)

**Comment by AnirG** [comment]
@jbarnoud sure, will do that.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/diffusionmap.py]:
```suggestion
    universe : `~MDAnalysis.core.universe.Universe` or `~MDAnalysis.core.groups.AtomGroup`
```

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (jbarnoud) [inline package/MDAnalysis/analysis/diffusionmap.py:344]: The change does not appear in the docstring just above.

**Review by jbarnoud — COMMENTED** [review]
I missed some tiny details during my previous readthrough. Once they are fixed, it should be good to go, assuming the other reviewers are satisfied.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: Apologies! I have updated the docstrings.

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Version 2.1.0 has now been released, you'll need to update your branch and add your changelog entry to 2.2.0 instead.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: Imports are usually grouped as "python built-ins", "external libraries", "internal library", see the PEP8 guide for more details: https://peps.python.org/pep-0008/#source-file-encoding

Can you move this to after line 159 please?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of extra things, you'll need to also resolve conflicts (which I think are related to the recent 2.1.0 -> 2.2.0 change)

File: package/CHANGELOG
Comment (AnirG) [reply package/CHANGELOG]: Sorry, didn't realize 2.1.0 was out already. I have revoked the changes and have updated on 2.2.0.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py:344]: Quick question here - what happens if you pass an UpdatingAtomGroup, does the code still work?

Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: yes, it works fine. PFA example code snippet I tried on.
[example.zip](https://github.com/MDAnalysis/mdanalysis/files/8323412/example.zip)

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Not sure why this white space change exists.
```suggestion
  * `packaging` is now a core dependency for MDAnalysis
```

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py:253]: Do add a versionchanged entry after line 247.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: Do add a versionchanged entry after line 314.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: Could you add in a test that uses an UpdatingAtomGroup?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Two documentation things and a test request.

**Review by richardjgowers — APPROVED** [review]

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: sure, will add a separate test for updating atomgroups.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: typo?

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: See here: https://mdanalysis--3541.org.readthedocs.build/en/3541/documentation_pages/analysis/diffusionmap.html#MDAnalysis.analysis.diffusionmap.DiffusionMap.transform

You'll need two empty lines for this to work as intended.
```suggestion

    .. versionchanged: 2.2.0
```

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: typo - extra space between class and classname
```suggestion
         :class:`DiffusionMap` now also accepts `AtomGroup`.
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of small extra things. Not sure why CI isn't triggering properly at the moment.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py]: yeah my bad :(

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py]: hi, I'm not very sure why it still doesn't work as intended even after giving 2 empty lines.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: Hi, I'm not very sure about what kind of test do we want here. Do I make a test out of the example.zip that I attached before and compare the DistMatrix constructed by updating atom group with just universe and selection string of the atom group?

Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py:344]: @IAlibay I did the above for updating AG checks. Please let me know if it is fine.

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: I'm not really understanding how this checks that an updating selection works, maybe I'm missing something obvious though.

Since universes are static, there shouldn't be fewer or more atoms that match "resname ALA" within the lifetime of a trajectory. Instead of you should aim to analyse something dynamic, like "ALA residues within a given distance of residue N".

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_diffusionmap.py
Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: yes my bad, I didn't give much attention to the atom selection. I was more concerned about if it worked when AG is an instance of UpdatingAtomGroup. I will update it such that selection is dynamic over all frames and the final selection is parsed over to `DistanceMatrix`. In my opinion, this test case is kinda redundant as we are checking if it's working for the final AG after all.

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: So the point of passing and UpdatingAtomGroup through and checking that everything still works as expected is because most Analysis methods in MDAnalysis use static defined numpy arrays. So if you end up with more or fewer entries than you'd expect, things can go wrong. Having a test that we can check works as expected is very important.

Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: @IAlibay I have a doubt here. Say, you make a selection as `selection = "around 5 resname ALA"`. For frame 1, let's assume this selection has 10 atoms, and for frame 50, it has 20 atoms. Do you expect RMSD(frame 1, frame 50) and RMSD(frame 50, frame 1) in the `DistanceMatrix` to be different since the selection is being updated dynamically due to the UpdatingAtomGroup? If this is what was implied here  https://github.com/MDAnalysis/mdanalysis/pull/3541/files/2799ad9cf2b9f12bf673a217985d781535f802a9#r831565393 , I believe it doesn't work this way. 
What can be done is package all these dynamic selections into one atomgroup and run our `DistanceMatrix` analysis on this AG (finally a static selection).

Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: @IAlibay Is my understanding correct?

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: @AnirG this is essentially the crux of what I was leading to with my original question on this, I was trying to make you think about the cases where a user passed in an UpdatingAtomgroup expecting that the there would be more / fewer atoms depending on the frames - would things work as expected?

Because of the way `metric` works, there is an expectation that the `i_ref` and `j_ref` arrays are of equal lengths (see https://github.com/MDAnalysis/mdanalysis/blob/009c467511210c442d15e4014f02e62e07dc9e85/package/MDAnalysis/analysis/diffusionmap.py#L185-L191), so as you observed above, this is unlikely to work.

It's fine in this case for the solution to just be "UpdatingAtomGroups are not accepted". Creating a complex solution involving the eventual generation of a static selection is probably not worth it.

The next steps here are then - what code changes are necessary here to avoid users passing UpdatinAtomGroups?

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: As @lilyminium just reminded me, https://github.com/MDAnalysis/mdanalysis/blob/e6f0fe1289d150747befd1f1773db5657760f2c5/package/MDAnalysis/analysis/diffusionmap.py#L256 essentially means that all UpdatingAtomGroups will be squashed to AtomGroups on class initilisation. Probably all you need here is to have a check on the input to see if it's an UpdatingAtomGroup and warn users that it is being converted to a static AtomGroup based on the current frame selection.

Comment (AnirG) [reply testsuite/MDAnalysisTests/analysis/test_diffusionmap.py]: @IAlibay thanks for clarifying, I'll make the necessary changes.

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/diffusionmap.py]: Please double check this, but I'm pretty sure it'll freeze it at the current frame, so if someone did:

```python
u.trajectory[10]
uag = u.select_atoms('resname HOH and within 4 resname LIG', updating=True)
dmatrix = DistanceMatrix(ag)
```

Then you're doing uag.select_atoms('all') on the current frame, rather than the first frame.

```suggestion
                    "will result in a static AtomGroup with the current frame "
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just the one last thing, otherwise everything looks good to me thanks!

**Comment by IAlibay** [comment]
Please do fix merge conflicts too.

**Comment by IAlibay** [comment]
@jbarnoud if you can also have a re-review that'd be great

File: package/MDAnalysis/analysis/diffusionmap.py
Comment (AnirG) [reply package/MDAnalysis/analysis/diffusionmap.py]: yes. you're right. It's the current frame.

**Review by IAlibay — APPROVED** [review]
lgtm, thanks for all the work @AnirG.

Just waiting on @jbarnoud to re-review.

**Review by jbarnoud — CHANGES_REQUESTED** [review]
Sorry, I'll ask for one last small test, and it is all good.

Users may provide both an AtomGroup and the select argument. From what I read, it should work well with your implementation, but I would like to see a test for it. That test does not have to be complicated, just to use both an AtomGroup and the select argument. I would write it as the same as `test_distvalues_ag_universe` but with 

```python
ag = u.select_atoms('protein'')
dist_ag = diffusionmap.DistanceMatrix(ag, select='backbone').run()
```

**Comment by jbarnoud** [comment]
Thank you for having taken care of the conflicts as well. I'll merge as soon as the CI is done and passing :)

**Comment by AnirG** [comment]
@IAlibay @jbarnoud thanks a lot for your patience and guidance :) felt quite satisfying starting out on this. Will hopefully solve more issues as I use the package more and get a better understanding over time.

**Review by jbarnoud — APPROVED** [review]
:tada:

**Comment by jbarnoud** [comment]
Congratulations! Thank you for addressing the issue :D

**Pull request merged**
